### PR TITLE
Simplify CloudEvent ctors and add dataSerializationType parameter

### LINKED ISSUE #17868: Consider exposing an optional Type parameter for EventGridEvent and CloudEvent constructors
Currently we serialize based on object.GetType(). This could lead to situations where we serialize derived properties that are unintended. For deserialization, we have a `GetData<T>` method where the type can be specified.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Adds an optional type parameter to EventGridEvent and CloudEvent constructors so users can specify the type to use when serializing. By default, GetType is still used. https://github.com/Azure/azure-sdk-for-net/issues/17868
- Collapse CloudEvent constructors into a single constructor with optional parameters for Data https://github.com/Azure/azure-sdk-for-net/issues/17421
- Get rid of special serialization behavior around BinaryData for both EventGridEvent and BinaryData. 
Previously, if a user passed BinaryData as the object to the CloudEvent constructor, we would always store the data in the DataBase64 field. This may not be the expected behavior if a user was working with JSON and expected the data to be delivered as JSON. For EventGridEvent, if BinaryData was used we would attempt to parse the BinaryData as JSON and, if successful, store as the Data object. If not, we would serialize the BinaryData as a string. Serializing as a string is incorrect as we could lose any data that couldn't be represented as a string. Parsing as JSON is bad for performance and error prone. To address all these issues, we will remove the special handling around BinaryData so that it would be treated like any other type that doesn't have settable properties. This means that if a user is attempting to pipe the BinaryData returned from GetData() from one topic to another, it won't work, and will instead end up sending an empty data object. We might want to consider adding an exception to explicitly stop people from doing this - currently it would just result in sending empty Data.

### CODE REVIEW

File: sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs [hunk 1]
```diff
@@ -7,9 +7,7 @@ public static partial class BinaryDataExtensions
     }
     public partial class CloudEvent
     {
-        public CloudEvent(string source, string type) { }
-        public CloudEvent(string source, string type, System.BinaryData data, string dataContentType = null) { }
-        public CloudEvent(string source, string type, object data, string dataContentType = null) { }
+        public CloudEvent(string source, string type, object data = null, string dataContentType = null, System.Type dataSerializationType = null) { }
```
Comment (pakrym) [inline sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: From some early UX studies, we had an idea that many optional parameters scare customers. That's why we opted for overloads in commonly used cases. This seems to do the oposite.

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: In general, users will be sending the data. By having a separate overload, users were not able to find the overload that contained the `data` parameter.

Comment (pakrym) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]:
``` C#
public CloudEvent(string source, string type, object data) 
public CloudEvent(string source, string type, object data = null, string dataContentType = null, System.Type dataSerializationType = null) 
```
??

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: I'm not sure if this is better. Now we introduce the scenario where most users will not set dataContentType.

Comment (pakrym) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: I think having a single `public CloudEvent(string source, string type, object data = null, string dataContentType = null, System.Type dataSerializationType = null) { }
` overload doesn't serves the "In general, users will want to set the data" goal.

Comment (pakrym) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: Can we make data nullable but not optional?

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: That is a possibility.

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: Let's see what others think, but you are proposing:
```cs 
public CloudEvent(string source, string type, object data, string dataContentType = null, System.Type dataSerializationType = null) 
```
And we just allow null data, right?

Comment (pakrym) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:10]: Or a pair of 
```
public CloudEvent(string source, string type, object data) 
public CloudEvent(string source, string type, object data, string dataContentType = null, System.Type dataSerializationType = null) 
```

**Review by ellismg — APPROVED** [review]

**Pull request merged**